src/coreclr/jit/inlinepolicy.cpp

```diff
@@ -1440,6 +1440,71 @@ void ExtendedDefaultPolicy::NoteDouble(InlineObservation obs, double value)
     m_ProfileFrequency = value;
 }
 
+//------------------------------------------------------------------------
+// BudgetCheck: see if this inline would exceed the current budget
+//
+// Returns:
+//   True if inline would exceed the budget.
+//
+bool ExtendedDefaultPolicy::BudgetCheck() const
+{
+    if (m_IsPrejitRoot)
+    {
+        // Only relevant if we're actually inlining.
+        return false;
+    }
+
+    // The strategy tracks the amout of inlining done so far, so it performs the actual check.
+    const bool overBudget = m_RootCompiler->m_inlineStrategy->BudgetCheck(m_CodeSize);
+
+    if (!overBudget)
+    {
+        // We're not over budget
+        return false;
+    }
+
+    assert(m_IsForceInlineKnown);
+    if (!m_IsForceInline)
+    {
+        // For over-budget case we only inspect forceinline
+        // Otherwise, report "over-budget"
+        return true;
+    }
+
+    if (m_CallsiteDepth == 1)
+    {
+        // Compatibility with DefaultPolicy:
+        // If the candidate is a forceinline and the callsite is
+        // not too deep, allow the inline even if it goes over budget.
+        JITDUMP("Allowing over-budget top-level forceinline\n");
+        return false;
+    }
+
+    assert(m_CallsiteDepth > 1);
+
+    // For this case we want to take number of foldable branches into account: JIT is able to elide them
+    // early in the importer so it won't have to waste time compiling them.
+
+    // Analyze potential benefits, we take into account:
+    //  * FoldableBranches
+    //  * FoldableSwitches (we usually elide more than in case of FoldableBranch, hence, multiply by 2)
+    // We assume that a single foldable branch is capable of removing 80 bytes of IL, e.g.:
+    //
+    //  if (typeof(TKey) == typeof(byte)) return (byte)(object)left < (byte)(object)right;
+    //
+    // This branch is recognized as foldable and removes 52 bytes of IL.
+    //
+    // In future, we plan to have a real estimation of what will be folded.
+    const double pros = (m_FoldableBranch + m_FoldableSwitch * 2.0) * 80;
+
+    // We take number of locals and IL size as a friction:
+    const double cons = this->m_CodeSize * (1.0 + (double)m_RootCompiler->lvaCount / JitConfig.JitMaxLocalsToTrack());
```

Comment:
Can this factor in locals we can potentially remove and don't have to track?

It's not uncommon, for example, for there to be 1 local per branch and so a given method might have 12 locals, but only 1 local will actually be live and need to be considered after the dead paths are dropped.

Reply:
For that we need to implement https://github.com/dotnet/runtime/issues/72740 first, but since we don't have it yet we have to rely on the total count.